java/velox4j/src/main/cpp/CMakeLists.txt

```diff
@@ -0,0 +1,129 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+cmake_minimum_required(VERSION 3.28)
+project(velox4jcpp)
+
+include(FetchContent)
+
+# Preconditions.
+if(NOT UNIX)
+  message(
+    FATAL_ERROR
+      "The project velox4j-cpp only supports UNIX-based systems at the moment.")
+endif()
+
+# Compile options.
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED TRUE)
+set(CMAKE_CXX_EXTENSIONS ON)
+set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+
+# Velox4J option definitions.
+option(VELOX4J_ENABLE_CCACHE "Use ccache if installed." ON)
+option(VELOX4J_BUILD_TESTING "Build Velox4J tests." ON)
+
+# Velox4J non-bool option definitions.
+set(VELOX4J_INSTALL_DESTINATION
+    "lib"
+    CACHE
+      STRING
+      "The directory Velox4J's libraries are installed. This is relative to the value of CMAKE_INSTALL_PREFIX."
+)
+
+# Setup JNI.
+find_package(Java REQUIRED)
+find_package(JNI REQUIRED)
+if(JNI_FOUND)
+  message(STATUS "JNI_INCLUDE_DIRS=${JNI_INCLUDE_DIRS}")
+  message(STATUS "JNI_LIBRARIES=${JNI_LIBRARIES}")
+endif()
+include(UseJava)
+
+# Set Velox options.
+set(VELOX_DEPENDENCY_SOURCE AUTO)
+set(folly_SOURCE BUNDLED) # Use velox-bundled folly because we have a custom
+                          # patch for folly setup.
+set(VELOX_ENABLE_CCACHE ON)
+set(VELOX_MONO_LIBRARY ON)
+set(VELOX_BUILD_SHARED ON)
+set(VELOX_BUILD_STATIC OFF)
+set(VELOX_ENABLE_ARROW ON)
+if(NOT VELOX4J_BUILD_TESTING)
+  set(VELOX_BUILD_TESTING OFF)
+endif()
+if(NOT VELOX4J_ENABLE_CCACHE)
+  set(VELOX_ENABLE_CCACHE OFF)
+endif()
+
+# Import Velox.
+set(VELOX_ROOT_DIR "${PROJECT_SOURCE_DIR}/../../../../../")
+FetchContent_Declare(
+  velox
+  DOWNLOAD_COMMAND
+    ln -sf ${VELOX_ROOT_DIR}/velox ${FETCHCONTENT_BASE_DIR}/velox-src/velox &&
+    ln -sf ${VELOX_ROOT_DIR}/scripts ${FETCHCONTENT_BASE_DIR}/velox-src/scripts
+    && ln -sf ${VELOX_ROOT_DIR}/CMake ${FETCHCONTENT_BASE_DIR}/velox-src/CMake
+    && ln -sf ${VELOX_ROOT_DIR}/CMakeLists.txt
+    ${FETCHCONTENT_BASE_DIR}/velox-src/CMakeLists.txt)
+FetchContent_MakeAvailable(velox)
```

Comment:
The core velox lib core is imported through `FetchContent`.

To avoid the recursive dependency `java -> velox -> java` mentioned in https://github.com/facebookincubator/velox/discussions/13044, we selectively create several symbolic links to these needed core velox files in java bindings' build dir:

- velox/velox
- velox/scripts
- velox/CMake
- velox/CMakeLists.txt

Symbolic links of these 4 items are enough to build the velox core lib as the dependency of java bindings' cpp code.

Reply:
> add_subdirectory(${VELOX_ROOT_DIR} ${VELOX_BINARY_DIR})

I once tried this but left out `${VELOX_BINARY_DIR}` so got some errors. Now it seems working fine with the binary dir set. Thanks for the suggestion. :+1: